src/Tools/DependencyManagement/ServiceProviders/QueueServiceProvider.php

```diff
@@ -0,0 +1,102 @@
+<?php
+/**
+ * QueueServiceProvider class file.
+ *
+ * @package Automattic\WooCommerce\Tools\DependencyManagement\ServiceProviders
+ */
+
+namespace Automattic\WooCommerce\Tools\DependencyManagement\ServiceProviders;
+
+use League\Container\ServiceProvider\AbstractServiceProvider;
+use \ReflectionClass;
+use \ReflectionException;
+use \WC_Action_Queue;
+use \WC_Queue_Interface;
+
+/**
+ * Service provider for WC_Queue_Interface.
+ */
+class QueueServiceProvider extends AbstractServiceProvider {
+
+	/**
+	 * The queue class that will be instantiated unless something different is specified via woocommerce_queue_class filter.
+	 */
+	const DEFAULT_QUEUE_CLASS = WC_Action_Queue::class;
+
+	/**
+	 * The classes/interfaces that are serviced by this service provider.
+	 *
+	 * @var array
+	 */
+	protected $provides = array(
+		WC_Queue_Interface::class,
+	);
+
+	/**
+	 * Register the function that will create the singleton instance of WC_Queue_Interface.
+	 */
+	public function register() {
+		$container = $this->getContainer();
+
+		$container->share(
+			WC_Queue_Interface::class,
+			function() {
+				return $this->get_queue_instance();
+			}
+		);
+	}
+
+	/**
+	 * Create a new instance of a class implementing WC_Queue_Interface.
+	 * It will be an instance of {DEFAULT_QUEUE_CLASS} unless something different is specified via woocommerce_queue_class filter.
+	 *
+	 * @return WC_Queue_Interface
+	 */
+	private function get_queue_instance() {
+		if ( ! did_action( 'plugins_loaded' ) ) {
+			$this->error_bad_timing();
+		}
+
+		$queue_class = apply_filters( 'woocommerce_queue_class', self::DEFAULT_QUEUE_CLASS );
```

Comment:
Since we're using this kind of filter in multiple places, perhaps we should make something more generic? For example, the session handler in this PR also has the same sort of filtering going on.

Reply:
What would you suggest? Something like a `woocommerce_get_service_class` filter that passes an interface name and a default class name, for example?

Reply:
In hindsight maybe a filter isn't a good choice in being forward-looking? I imagine code that relies on the interface will throw an exception if it is incompatible anyway so we might just want to expose some way for the container to be altered by plugins. At least that way they're explicitly depending on the interface, but in general, they probably should avoid this because of API changes causing fatal exceptions.

Reply:
If plugins have access to the container (and they should) then they can register whatever they want, for good and for bad. That's why I was initially using an `ObjectContainer` that didn't expose the container directly, only its `get` method. 🙂 Maybe we could somehow override its `add` method to disallow re-registering internal classes (except some selected interfaces) after the initial setup performed in core?

Reply:
Discussed in Slack, nobody should be re-registering but we should provide a means to replace concrete implementations in the underlying container.